File: julia/31_statevector2mps.jl
#!/usr/bin/env julia
"""
Generate MPS from statevector
"""

using LinearAlgebra

function main()
    cutoff = 1e-10

    # Bell state
    println("Bell state:")
    v = [1.0, 0.0, 0.0, 1.0] ./ sqrt(2)

    v = permutedims(reshape(v, 2, 2))
    F = svd(v; full=false)
    println("singular values: ", F.S)
    rank_new = sum(F.S .> cutoff * F.S[1])
    U = @view F.U[:, 1:rank_new]
    S = Diagonal(F.S[1:rank_new])
    Vt = @view(F.V[:, 1:rank_new])'
    v = S * Vt
    println("tensors ", Any[U, v], ")\n")

    # GHZ state
    n = 16
    println("n=$n GHZ state:")
    v = zeros(Float64, 2^n)
    v[1] = 1 / sqrt(2)
    v[end] = 1 / sqrt(2)

    mps = Array{Float64}[]
    rank = 1
    for i in 0:(n-2)
        v = reshape(v, rank * 2, :)
        F = svd(v; full=false)
        println("$(i): singular values: ", F.S)
        rank_new = sum(F.S .> cutoff * F.S[1])
        U = @view F.U[:, 1:rank_new]
        S = Diagonal(F.S[1:rank_new])
        Vt = @view(F.V[:, 1:rank_new])'
        if i > 0
            U = reshape(U, rank, 2, rank_new)
        end
        push!(mps, U)
        v = S * Vt
        rank = rank_new
    end
    v = reshape(v, rank, 2)
    push!(mps, v)
    println("tensors: ", mps, "\n")

    # random state
    n = 16
    println("n=$n random state:")
    v = rand(2^n)
    v = v / norm(v)

    mps = Array{Float64}[]
    rank = 1
    for i in 0:(n-2)
        v = reshape(v, rank * 2, :)
        F = svd(v; full=false)
        println("$(i): singular values: ", F.S)
        rank_new = sum(F.S .> cutoff * F.S[1])
        U = @view F.U[:, 1:rank_new]
        S = Diagonal(F.S[1:rank_new])
        Vt = @view(F.V[:, 1:rank_new])'
        if i > 0
            U = reshape(U, rank, 2, rank_new)
        end
        push!(mps, U)
        v = S * Vt
        rank = rank_new
    end
    v = reshape(v, rank, 2)
    push!(mps, v)
    println("virtual bond dimensions:")
    for i in 2:n
        print("$(i-1): $(size(mps[i])[1])\n")
    end
end

main()

File: python/31_statevector2mps.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Generate MPS from statevector
"""

import numpy as np


def main():
    cutoff = 1e-10

    # Bell state
    print("Bell state:")
    v = np.array([1, 0, 0, 1]) / np.sqrt(2)

    v = v.reshape((2, 2))
    U, S, Vt = np.linalg.svd(v, full_matrices=False)
    print(f"singular values: {S}")
    S = S[S > cutoff * S[0]]
    rank_new = len(S)
    U = U[:, :rank_new]
    Vt = Vt[:rank_new, :]
    v = np.dot(np.diag(S), Vt)
    print(f"tensors {[U, v]})\n")

    # GHZ state
    n = 16
    print(f"n={n} GHZ state:")
    v = np.zeros(2**n)
    v[0] = 1 / np.sqrt(2)
    v[-1] = 1 / np.sqrt(2)

    mps = []
    rank = 1
    for i in range(0, n - 1):
        v = v.reshape((rank * 2, -1))
        U, S, Vt = np.linalg.svd(v, full_matrices=False)
        print(f"{i}: singular values: {S}")
        S = S[S > cutoff * S[0]]
        rank_new = len(S)
        U = U[:, :rank_new]
        Vt = Vt[:rank_new, :]
        if i > 0:
            U = U.reshape((rank, 2, rank_new))
        mps.append(U)
        v = np.dot(np.diag(S), Vt)
        rank = rank_new
    v = v.reshape((rank, 2))
    mps.append(v)
    print(f"tensors: {mps}\n")

    # random state
    n = 16
    print(f"n={n} random state:")
    v = np.random.random(2**n)
    v = v / np.linalg.norm(v)

    mps = []
    rank = 1
    for i in range(0, n - 1):
        v = v.reshape((rank * 2, -1))
        U, S, Vt = np.linalg.svd(v, full_matrices=False)
        print(f"{i}: singular values: {S}")
        S = S[S > cutoff * S[0]]
        rank_new = len(S)
        U = U[:, :rank_new]
        Vt = Vt[:rank_new, :]
        if i > 0:
            U = U.reshape((rank, 2, rank_new))
        mps.append(U)
        v = np.dot(np.diag(S), Vt)
        rank = rank_new
    v = v.reshape((rank, 2))
    mps.append(v)
    print("virtual bond dimensions:")
    for i in range(1, n):
        print(f"{i - 1}: {mps[i].shape[0]}")


if __name__ == "__main__":
    main()
